### ISSUE TITLE
SOCKS Proxy Support

### ISSUE THREAD

**Issue by Ohge** [feross/webtorrent#807]
Can you please add SOCKS proxy support? I am working on a crawler that parses metadata, but I have received two fraudulent DMCA requests this week even though I never shared a single byte of the actual files (destroys torrent on metadata). The only way to avoid having my internet shut off while using this software is to push all the traffic through a proxy or VPN so my lSP is hidden. My primary concern is support in node.js, but if it can be supported client side as well that would help people who use proxies in their browsers for similar reasons.

P.S. This is great software, and I intend to add a bounty to this feature request.

**Comment by feross** [comment]
There's no way to add proxy support to the browser version of WebTorrent. That will always use your browser/OS proxy settings because JavaScript in a web page can't control that, for security reasons.

As for SOCKS support, it seems there's two possible ways to do it: proxy only connections to trackers, or proxy all peer connections. `libtorrent` [has support](http://libtorrent.org/manual.html#proxy-settings) for both options.

**Comment by feross** [comment]
This package actually looks quite good: https://www.npmjs.com/package/socks

**Comment by Ohge** [comment]
I've used libtorrent before for python stuff, but I really like your node implementation because I can just pass the magnet and file index to the client and they can stream the file themselves with zero client side dependencies beyond the browser. I'm not in any rush or demanding anything, as I have already moved my code to a VM behind a VPN, but I thought it would be much easier for most developers (myself included) if they could just pass in the proxy/auth details in the opts arguments when creating the instance, rather than putting all the traffic on the machine behind a VPN.

In any event this is some extremely cool software, and I hope you'll keep improving it, and adding features. Proxy or not you deserve the bounty. Thank you for sharing!

**Comment by feross** [comment]
Yeah, that's totally fair! I think this wouldn't be that hard to support, so it'll probably happen.

The `socks` dependency seems pretty lightweight. Or even better, we can avoid the dependency and just let the user pass in an instance of that or something.

Will look at this when I have some time. PR welcome too!

**Comment by feross** [comment]
Also, for others coming along later, this is the bounty that @Ohge is referring to: https://www.bountysource.com/issues/34324857-socks-proxy-support

**Comment by romaincointepas** [comment]
Definitely interested in this as well, just added $50 to the bounty!

**Comment by yciabaud** [comment]
I have some difficulties to set a SOCKS proxy for UDP up to test my PR feross/bittorrent-tracker#159 to enable socks proxy on tracker requests.
I tried dante server but I could not achieve to connect to it.

It must be a configuration issue but I could not fix it...

Does anyone can help me with this by testing or explain how to setup UDP forwarding?

**Comment by romaincointepas** [comment]
@yciabaud You could test using https://www.privateinternetaccess.com (free trial I think), they offer SOCKS5 proxy specifically for torrenting (so it should support UDP/TCP, also good to test authentification).

**Comment by yciabaud** [comment]
@romaincointepas Unfortunately, there is no free trial and the free proxies I found only allow http browsing.

It would have been easier to setup a local proxy, I will give it another try...

**Comment by romaincointepas** [comment]
@yciabaud I just emailed you the PIA SOCKS5 proxy (UDP/TCP and made for torrenting) with my personal credentials, good luck!

**Comment by yciabaud** [comment]
Thank you for letting me use this, I tried yesterday and I achieved to use the proxy for HTTP and Websocket trackers. Unfortunately the connection is not established on UDP trackers.

Digging into this, I will keep you in touch.

**Comment by yciabaud** [comment]
Little update, I got the negotiation for UDP relay working but the messages are still not forwarded to the destination.

Should work on this soon.

**Comment by romaincointepas** [comment]
@yciabaud Looking forward, good luck!

**Comment by dcposch** [comment]
@yciabaud curious, what all gets proxied?

WebTorrent talks to the network it lots of ways. So far I've found:
1. Connecting to traditional trackers over UDP
2. Connecting to traditional trackers over HTTP/S
3. Connecting to WebTorrent trackers via websockets
4. Connecting to peers by making an outgoing TCP connection
5. Connecting to peers by listening on a port, accepting incoming TCP connections
6. Connecting to peers over WebRTC
7. Connecting to web seeds over HTTP/S

The connections are made from a few different modules: webtorrent, simple-get, and torrent-discovery.

Do we need to proxy all of those?

**Comment by yciabaud** [comment]
You're right @dcposch, I am currently working on bittorrent-tracker to cover:
> 1. Connecting to traditional trackers over UDP - In progress
> 2. Connecting to traditional trackers over HTTP/S - Done
> 3. Connecting to WebTorrent trackers via websockets - Done

But then we should add proxying connections to peers as well:
> 4. Connecting to peers by making an outgoing TCP connection
> 7. Connecting to web seeds over HTTP/S

The last cases are a bit different, I don't know if we need to manage proxying incoming connections and webrtc, some proxies can open a port and forward incoming messages to you but I don't think other clients has implemented it. For webrtc, TURN/STUN can be used but it may be better to use the configured proxy.
> 5. Connecting to peers by listening on a port, accepting incoming TCP connections
> 6. Connecting to peers over WebRTC

Finally I would add a last networking issue:
8. Connect to TURN servers

In that case, proxy may be needed to access the servers.

Let's start with tracker and peers support for outgoing connections and then discuss on the other cases.

**Comment by romaincointepas** [comment]
@yciabaud @dcposch Any update on this? I would be more than happy to bump the bounty quite a bit if that helps.

**Comment by yciabaud** [comment]
No update on my side, I was quite stuck on UDP tracker proxy, I should have another look this week.

The code looks good but I did achieve to connect to a UDP tracker yet...

**Comment by yciabaud** [comment]
I cannot get vuze to work with UDP, tried @romaincointepas proxy and a local 3proxy instance and I have the same behavior... UDP negociation works but then the request times out...

Can somebody test it for me? Maybe my ISP is locking some kind of traffic :-1:

**Comment by romaincointepas** [comment]
Most SOCKS5 proxy servers out there are not compatible with UDP trackers (3proxy being one of them). As far as I know (I've tried a lot of them), only dante-server works.

See this [comment](https://github.com/arvidn/libtorrent/issues/233#issuecomment-150838354) from the creator of libtorrent:
> My experience with udp over socks5 is that very few proxies implement it correctly or at all. As far as I can recall, I've only seen one that does (dante).

My uneducated guess is that it may have something to do with http://www.bittorrent.org/beps/bep_0015.html (**UDP connections / spoofing**):
> In the ideal case, only 2 packets would be necessary. However, it is possible to spoof the source address of a UDP packet. The tracker has to ensure this doesn't occur, so it calculates a value (connection_id) and sends it to the client. **If the client spoofed it's source address, it won't receive this value (unless it's sniffing the network)**. The connection_id will then be send to the tracker again in packet 3. The tracker verifies the connection_id and ignores the request if it doesn't match. Connection IDs should not be guessable by the client. This is comparable to a TCP handshake and a syn cookie like approach can be used to storing the connection IDs on the tracker side. A connection ID can be used for multiple requests. A client can use a connection ID until one minute after it has received it. Trackers should accept the connection ID until two minutes after it has been send.

PIA, Torguard and other VPN providers all seem to use dante-server, and I can get a proxied setup working with http://www.qbittorrent.org + PIA SOCKS5 proxy (the one I sent you by email @yciabaud, can you try with qBittorrent?).

Also, could http://www.bittorrent.org/beps/bep_0015.html (UDP connections / spoofing) mean that there is some additional trickery needed client-side for UDP trackers to work?

**Comment by yciabaud** [comment]
OK got it working now, at first I set a local dante server to have logs and then I checked with your proxy server.

So here is the new status (when PR feross/bittorrent-tracker#157 is merged):

- [x] Connecting to traditional trackers over UDP
- [x] Connecting to traditional trackers over HTTP/S
- [x] Connecting to WebTorrent trackers via websockets
- [ ] Connecting to web seeds over HTTP/S
- [ ] Connecting to peers by making an outgoing TCP connection
- [ ] Connecting to peers by listening on a port, accepting incoming TCP connections
- [ ] Connecting to peers over WebRTC
- [ ] Connecting to TURN servers

Connecting to peers should not be too difficult for HTTP and TCP. WebRTC and TURN will not be straightforward so we can look at this later.

**Comment by romaincointepas** [comment]
@yciabaud Great!

What happens if the SOCKS5 server does not respond (either right away or later on)? Does bittorrent-tracker falls back on direct connection or times out? (for each case)

**Comment by yciabaud** [comment]
No fallback, the request will time out and we will wait the next announce interval to try again.

**Comment by yciabaud** [comment]
Worked on proxying peer connections today, needs more testing but webseeds and tcp should be ok.

**Comment by Miserlou** [comment]
Excellent news!

**Comment by samholmes** [comment]
What's the status on this?

**Comment by erlinux** [comment]
i need it.

**Comment by Jad3z** [comment]
Any updates? This is kinda big :p

**Comment by fredsif** [comment]
Add 25$ to the bounty.
Will keep adding every month until it's done.

**Comment by actuallymentor** [comment]
If anyone is working on this and needs a SOCKS5 to connect to, feel free to use `socks.generousvpn.com` on port `1080`. It uses the aforementioned Dante proxy package.

*Disclosure: I run generousvpn.*

**Comment by nake89** [comment]
Any new developments regarding this?

**Comment by Fenny** [comment]
I think webtorrent had his prime, specially since IPFS is growing rapidly.

**Comment by balupton** [comment]
Couldn't one setup a docker vm to run webtorrent and a vpn client on the same box to achieve this?

**Comment by Ohge** [comment]
> Couldn't one setup a docker vm to run webtorrent and a vpn client on the same box to achieve this?

That was essentially the work around I was using back in 2016 when I first opened the issue. It does work that way, but my intention was to use webtorrent in another piece of software (like sickbeard, couchpotato, headphones, etc which all support proxies). Think local network app that indexes popular video files in the DHT and makes them available for streaming/download. Unfortunately, because I couldn't add proxy support (like similar apps) for the end users I stopped working on the project hoping support would be added later and I could pick it up again.

**Comment by resession** [comment]
i would love this, especially for the dht.

**Comment by erlinux** [comment]
Any new developments regarding this?

**Comment by ralyodio** [comment]
also wondering i'm using the node.js module on the server, does socks5 proxy work yet?

**Comment by QuixThe2nd** [comment]
I've just read over #874, it seems a lot of progress has been made. What is the reason why it was never merged? I'd be happy to contribute towards whatever is needed to getting SOCKS proxy support implemented.

**Comment by LinneRELLa** [comment]
Me too, I don't know why the development progress for this feature stopped here.
